pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1184,168 +898,127 @@ protected IdealState ensureAllPartitionsConsuming(final TableConfig tableConfig,
         // i.e. after updating old segment metadata (old segment metadata state = DONE)
         // and creating new segment metadata (new segment metadata state = IN_PROGRESS),
         // but before updating ideal state (new segment ideal missing from ideal state)
-        if (isTooSoonToCorrect(tableNameWithType, segmentId, now)) {
-          continue;
-        }
-
-        Preconditions
-            .checkArgument(latestMetadata.getStatus().equals(CommonConstants.Segment.Realtime.Status.IN_PROGRESS));
-        LOGGER.info("{}:Repairing segment for partition {}. Segment {} not found in idealstate", tableNameWithType,
-            partition, segmentId);
-
-        // If there was a prev segment in the same partition, then we need to fix it to be ONLINE.
-        LLCRealtimeSegmentZKMetadata secondLatestMetadata = latestMetadataArray[1];
-        if (secondLatestMetadata == null && skipNewPartitions) {
+        if (!isExceededMaxSegmentCompletionTime(realtimeTableName, latestSegmentName, currentTimeMs)) {
           continue;
         }
-        if (secondLatestMetadata != null) {
-          onlineSegments.add(secondLatestMetadata.getSegmentName());
+        LOGGER.info("Repairing segment: {} which has segment ZK metadata but does not exist in IdealState",
+            latestSegmentName);
+
+        if (latestSegmentZKMetadata.getStatus() == Status.IN_PROGRESS) {
```

Comment:
previously, we were getting metadata for the last 2 segments. Just curious, why did you change that to just latest segment, and instead opt to get the second last metadata here, by scanning all the instanceStatesMap for every partition?

Reply:
We cannot rely on the segment ZK metadata because previous segment might get all replicas OFFLINE, and we need to skip such segments (covered by the tests).
There is no ZK fetch for the instanceStatesMap scan, so it is actually much cheaper than fetching one more segment ZK metadata for each partition.